src/app/util/ember-compatibility-functions.cpp

```diff
@@ -575,36 +580,66 @@ CHIP_ERROR WriteSingleClusterData(ClusterInfo & aClusterInfo, TLV::TLVReader & a
     attributePathParams.mClusterId   = aClusterInfo.mClusterId;
     attributePathParams.mAttributeId = aClusterInfo.mAttributeId;
 
+    Protocols::InteractionModel::Status status = Protocols::InteractionModel::Status::Failure;
+
     // Named aPath for now to reduce the amount of code change that needs to
     // happen when the above TODO is resolved.
     ConcreteDataAttributePath aPath(aClusterInfo.mEndpointId, aClusterInfo.mClusterId, aClusterInfo.mAttributeId);
-    const EmberAfAttributeMetadata * attributeMetadata =
-        emberAfLocateAttributeMetadata(aPath.mEndpointId, aPath.mClusterId, aPath.mAttributeId, CLUSTER_MASK_SERVER, 0);
-
-    if (attributeMetadata == nullptr)
     {
-        return apWriteHandler->AddStatus(attributePathParams, Protocols::InteractionModel::Status::UnsupportedAttribute);
-    }
+        AttributeAccessInterface * attrOverride = findAttributeAccessOverride(aClusterInfo.mEndpointId, aClusterInfo.mClusterId);
+        if (attrOverride != nullptr)
+        {
+            AttributeValueDecoder valueDecoder(aReader, apWriteHandler->GetAccessingFabricIndex());
+            ReturnErrorOnFailure(attrOverride->Write(aPath, valueDecoder));
 
-    if (attributeMetadata->IsReadOnly())
-    {
-        return apWriteHandler->AddStatus(attributePathParams, Protocols::InteractionModel::Status::UnsupportedWrite);
+            if (valueDecoder.TriedDecode())
+            {
+                status = Protocols::InteractionModel::Status::Success;
+                ExitNow();
```

Comment:
please extract this in some other way, maybe a separate method that returns the 'status' and then one that does the apWriteHandler->AddStatus call on the returned  status.

The `goto exit` pattern makes the code harder to follow since whenever seeing an exit I have to go and see what the exit block does.

Reply:
Offline conversation - code was messy to start with and likely refactors need to be owned separately. Will let updates to be a separate PR.